server/src/com/thoughtworks/go/server/service/EnvironmentConfigService.java

```diff
@@ -145,6 +136,24 @@ public EnvironmentConfig named(String environmentName) throws NoSuchEnvironmentE
         return environments.named(new CaseInsensitiveString(environmentName));
     }
 
+    public EnvironmentConfig getEnvironmentConfig(String environmentName) throws NoSuchEnvironmentException {
+        if(environments.hasEnvironmentNamed(new CaseInsensitiveString(environmentName)))
+            return environments.named(new CaseInsensitiveString(environmentName));
+        return null;
+    }
+
+    public void deleteEnvironment(String environmentName, HttpLocalizedOperationResult result) throws NoSuchEnvironmentException {
+        CruiseConfig config = goConfigService.getConfigForEditing();
+        EnvironmentsConfig environments = config.getEnvironments();
+        if (environments.remove(getEnvironmentConfig(environmentName))) {
+            config.setEnvironments(environments);
```

Comment:
Also any edits to an existing environment by another thread may be lost here because you remember the "old state" of ALL environments.